src/main/java/org/openstreetmap/atlas/geography/sharding/AppendCountsForAllZoom.java

```diff
@@ -0,0 +1,179 @@
+     *
+     * @param zoomLayerToGenerate
+     *            the zoom layer for which to generate counts
+     * @param countsAtHigherZoom
+     *            HashMap containing counts for all {@link SlippyTile} in zoomLayerToGenerate+1
+     * @return a HashMap containing counts for all (@link SlippyTile} in zoomLayerToGenerate
+     */
+    private HashMap<SlippyTile, Long> calculateTileCountsForZoom(final int zoomLayerToGenerate,
+            final HashMap<SlippyTile, Long> countsAtHigherZoom)
+    {
+        long count = 0;
+        long tilesCalculated = 0;
+        final HashMap<SlippyTile, Long> countsAtThisZoom = new HashMap<>();
+        for (int x = 0; x < Math.pow(2, zoomLayerToGenerate + 1); x += 2)
+        {
+            for (int y = 0; y < Math.pow(2, zoomLayerToGenerate + 1); y += 2)
+            {
+                count = 0;
+                // top left
+                count += countsAtHigherZoom
+                        .getOrDefault(new SlippyTile(x, y, zoomLayerToGenerate + 1), (long) 0);
+                // top right
+                count += countsAtHigherZoom
+                        .getOrDefault(new SlippyTile(x + 1, y, zoomLayerToGenerate + 1), (long) 0);
+                // bottom left
+                count += countsAtHigherZoom
+                        .getOrDefault(new SlippyTile(x, y + 1, zoomLayerToGenerate + 1), (long) 0);
+                // bottom right
+                count += countsAtHigherZoom.getOrDefault(
+                        new SlippyTile(x + 1, y + 1, zoomLayerToGenerate + 1), (long) 0);
+                if (count != 0)
+                {
+                    countsAtThisZoom.put(new SlippyTile(x / 2, y / 2, zoomLayerToGenerate), count);
+                }
+                if (++tilesCalculated % READER_REPORT_FREQUENCY == 0)
+                {
+                    logger.info("Calculated {} zoom level {} tiles.", tilesCalculated,
+                            zoomLayerToGenerate);
+                }
+            }
+        }
+        return countsAtThisZoom;
+    }
+
+    /**
+     * Writes to file all the {@link SlippyTile} feature counts at a certain zoom level.
+     *
+     * @param countsAtZoom
+     *            HashMap containing the feature counts at that zoom level.
+     * @param bufferedWriter
+     *            The writer to write to.
+     */
+    private void writeToFile(final HashMap<SlippyTile, Long> countsAtZoom,
+            final BufferedWriter bufferedWriter)
+    {
+        final String lineSeparator = System.getProperty("line.separator");
+        countsAtZoom.forEach((tile, count) ->
+        {
+            try
+            {
+                bufferedWriter.write(tile.getName() + "," + count);
+                bufferedWriter.write(lineSeparator);
+            }
+            catch (final IOException e)
+            {
+                logger.error("Error writing converted data to {}", bufferedWriter);
```

Comment:
Does it make sense to throw here? What happens if we don't write for a specific zoom?

Reply:
You're right, it does. If we don't write for a specific zoom the output of this job will cause `DynamicTileSharding` to fail.